Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore

cdumez opened the pull request:
#### 9cce1d93832cf061ec64a78217e487a71634abc8
<pre>
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=285251">https://bugs.webkit.org/show_bug.cgi?id=285251</a>

Reviewed by Darin Adler.

* Source/WebCore/Modules/mediastream/RTCPeerConnection.cpp:
(WebCore::certificateTypeFromAlgorithmIdentifier):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpTransformableFrame.cpp:
(WebCore::LibWebRTCRtpTransformableFrame::data const):
* Source/WebCore/animation/BlendingKeyframes.cpp:
(WebCore::BlendingKeyframes::operator== const):
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::writeLittleEndian):
(WebCore::CloneDeserializer::deserializeString):
(WebCore::CloneDeserializer::readLittleEndian):
(WebCore::CloneDeserializer::readString):
(WebCore::CloneDeserializer::readStringData):
(WebCore::CloneDeserializer::readTag):
(WebCore::CloneDeserializer::readArrayBufferViewSubtag):
(WebCore::CloneDeserializer::readArrayBufferImpl):
(WebCore::CloneDeserializer::readResizableNonSharedArrayBuffer):
(WebCore::CloneDeserializer::read):
(WebCore::CloneDeserializer::readTerminal):
(WebCore::CloneDeserializer::consumeCollectionDataTerminationIfPossible):
* Source/WebCore/html/FTPDirectoryDocument.cpp:
(WebCore::FTPDirectoryDocumentParser::parseAndAppendOneLine):
* Source/WebCore/html/canvas/OESDrawBuffersIndexed.cpp:
* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
* Source/WebCore/html/canvas/WebGLDrawBuffers.cpp:
(WebCore::WebGLDrawBuffers::drawBuffersWEBGL):
* Source/WebCore/html/canvas/WebGLMultiDraw.cpp:
(WebCore::WebGLMultiDraw::multiDrawArraysWEBGL):
(WebCore::WebGLMultiDraw::multiDrawArraysInstancedWEBGL):
(WebCore::WebGLMultiDraw::multiDrawElementsWEBGL):
(WebCore::WebGLMultiDraw::multiDrawElementsInstancedWEBGL):
* Source/WebCore/html/canvas/WebGLMultiDrawInstancedBaseVertexBaseInstance.cpp:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
(WebCore::WebGLRenderingContextBase::TypedList::span const):
* Source/WebCore/loader/FTPDirectoryParser.cpp:
(WebCore::parseOneFTPLine):
* Source/WebCore/loader/FTPDirectoryParser.h:
(WebCore::ListResult::clear):
(WebCore::ListResult::filenameSpan const): Deleted.
(WebCore::ListResult::linknameSpan const): Deleted.
* Source/WebCore/loader/LinkHeader.cpp:
(WebCore::findURLBoundaries):
(WebCore::parseParameterName):
(WebCore::skipQuotesIfNeeded):
* Source/WebCore/loader/appcache/ApplicationCacheManifestParser.cpp:
* Source/WebCore/loader/archive/ArchiveResource.cpp:
(WebCore::ArchiveResource::saveToDisk):
* Source/WebCore/mathml/MathMLOperatorDictionary.cpp:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleTouchEvent):
* Source/WebCore/page/cocoa/ResourceUsageThreadCocoa.mm:
(WebCore::task_threads_span):
(WebCore::threadInfos):
* Source/WebCore/page/mac/ServicesOverlayController.mm:
(WebCore::expandForGap):
(WebCore::compactRectsWithGapRects):

Canonical link: <a href="https://commits.webkit.org/288347@main">https://commits.webkit.org/288347@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9eee5219061dc5fbf122084a561e7d4054b187fc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82684 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2335 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36844 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87804 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33742 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84778 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2407 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10246 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64426 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22195 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85741 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/1755 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75211 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44701 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1650 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29397 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32776 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72926 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30057 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89170 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9977 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7228 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72842 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10205 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71020 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72056 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16205 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1366 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12813 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9930 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15451 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9804 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13269 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11573 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -3378,39 +3373,42 @@ class CloneDeserializer : public CloneBase {
-            ptr += sizeof(T);
+        if (sizeof(T) == 1) {
+            value = span[0];
+            span = span.subspan(1);
```
Why do we need a special case here for size 1? If we do need it maybe `if constexpr` is better?

darinadler commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -3378,39 +3373,42 @@ class CloneDeserializer : public CloneBase {
-        if (sizeof(T) == 1)
-            value = *ptr++;
-        else {
+        if (sizeof(T) == 1) {
```
Maybe `if constexpr`?

darinadler commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -3378,39 +3373,42 @@ class CloneDeserializer : public CloneBase {
-            for (unsigned i = 0; i < sizeof(T); i++)
-                value += ((T)*ptr++) << (i * 8);
+            for (size_t i = 0; i < sizeof(T); ++i)
+                value += ((T)span[i]) << (i * 8);
```
Does this need to be a C-style cast? Maybe we can assign to a local variable of type `T` and avoid a cast. Or maybe `static_cast` if that won't work.

cdumez replied:
Looks like static_cast works.

darinadler commented on Source/WebCore/loader/FTPDirectoryParser.cpp:
```diff
@@ -168,8 +168,7 @@ FTPEntryType parseOneFTPLine(const char* line, ListState& state, ListResult& res
                         p = &(line[lineLenSansWsp]);
-                        result.filename = tokens[1];
-                        result.filenameLength = p - tokens[1];
+                        result.filename = std::span(tokens[1], p - tokens[1]);
```
Isn't there a `span` constructor that takes two pointers? Or am I imagining things?

cdumez replied:
Oh, you're right! I'll update.

darinadler commented on Source/WebCore/loader/FTPDirectoryParser.cpp:
```diff
@@ -769,9 +759,7 @@ FTPEntryType parseOneFTPLine(const char* line, ListState& state, ListResult& res
-                result.filename = &p[53 - 18];
-                result.filenameLength = (&(line[lineLenSansWsp]))
-                    - (result.filename);
+                result.filename = std::span(&p[53 - 18], &(line[lineLenSansWsp])- result.filename.data());
```
Missing space before `-` here.

darinadler commented on Source/WebCore/loader/FTPDirectoryParser.cpp:
```diff
@@ -970,35 +958,29 @@ FTPEntryType parseOneFTPLine(const char* line, ListState& state, ListResult& res
-                        result.linkname = result.filename + (result.filenameLength - fileSize);
-                        result.linknameLength = (&(line[linelen])) - (result.linkname);
-                        result.filenameLength -= fileSize + 4;
+                    if (result.filename.size() > (fileSize + 4) && !strncmp(result.filename.data() + result.filename.size() - fileSize - 4, " -> ", 4)) {
```
The `strncmp` call here doesn't seem super safe.

cdumez replied:
Yes, this whole filed is currently marked as `WTF_ALLOW_UNSAFE_BUFFER_USAGE`. It will need quite a bit of work to make it buffer-safe, though I can look into this particular strncmp() call.

darinadler reviewed (approved):
Seems in future like we could benefit from functions that extract values from the beginning of a `span` that combine the casting of the data extracted or the `memcpy` with the `subspan` call so we don't have to say the size twice and possibly get it wrong. At least 10 places that could use that; maybe 5 I noticed in this PR. Maybe somehow aligned with `StringParsingBuffer` functions that also "parse" things off the start of a range?

cdumez committed:
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore
https://bugs.webkit.org/show_bug.cgi?id=285251

Reviewed by Darin Adler.

* Source/WebCore/Modules/mediastream/RTCPeerConnection.cpp:
(WebCore::certificateTypeFromAlgorithmIdentifier):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpTransformableFrame.cpp:
(WebCore::LibWebRTCRtpTransformableFrame::data const):
* Source/WebCore/animation/BlendingKeyframes.cpp:
(WebCore::BlendingKeyframes::operator== const):
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::writeLittleEndian):
(WebCore::CloneDeserializer::deserializeString):
(WebCore::CloneDeserializer::readLittleEndian):
(WebCore::CloneDeserializer::readString):
(WebCore::CloneDeserializer::readStringData):
(WebCore::CloneDeserializer::readTag):
(WebCore::CloneDeserializer::readArrayBufferViewSubtag):
(WebCore::CloneDeserializer::readArrayBufferImpl):
(WebCore::CloneDeserializer::readResizableNonSharedArrayBuffer):
(WebCore::CloneDeserializer::read):
(WebCore::CloneDeserializer::readTerminal):
(WebCore::CloneDeserializer::consumeCollectionDataTerminationIfPossible):
* Source/WebCore/html/FTPDirectoryDocument.cpp:
(WebCore::FTPDirectoryDocumentParser::parseAndAppendOneLine):
* Source/WebCore/html/canvas/OESDrawBuffersIndexed.cpp:
* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
* Source/WebCore/html/canvas/WebGLDrawBuffers.cpp:
(WebCore::WebGLDrawBuffers::drawBuffersWEBGL):
* Source/WebCore/html/canvas/WebGLMultiDraw.cpp:
(WebCore::WebGLMultiDraw::multiDrawArraysWEBGL):
(WebCore::WebGLMultiDraw::multiDrawArraysInstancedWEBGL):
(WebCore::WebGLMultiDraw::multiDrawElementsWEBGL):
(WebCore::WebGLMultiDraw::multiDrawElementsInstancedWEBGL):
* Source/WebCore/html/canvas/WebGLMultiDrawInstancedBaseVertexBaseInstance.cpp:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
(WebCore::WebGLRenderingContextBase::TypedList::span const):
* Source/WebCore/loader/FTPDirectoryParser.cpp:
(WebCore::parseOneFTPLine):
* Source/WebCore/loader/FTPDirectoryParser.h:
(WebCore::ListResult::clear):
(WebCore::ListResult::filenameSpan const): Deleted.
(WebCore::ListResult::linknameSpan const): Deleted.
* Source/WebCore/loader/LinkHeader.cpp:
(WebCore::findURLBoundaries):
(WebCore::parseParameterName):
(WebCore::skipQuotesIfNeeded):
* Source/WebCore/loader/appcache/ApplicationCacheManifestParser.cpp:
* Source/WebCore/loader/archive/ArchiveResource.cpp:
(WebCore::ArchiveResource::saveToDisk):
* Source/WebCore/mathml/MathMLOperatorDictionary.cpp:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleTouchEvent):
* Source/WebCore/page/cocoa/ResourceUsageThreadCocoa.mm:
(WebCore::task_threads_span):
(WebCore::threadInfos):
* Source/WebCore/page/mac/ServicesOverlayController.mm:
(WebCore::expandForGap):
(WebCore::compactRectsWithGapRects):

Canonical link: https://commits.webkit.org/288347@main
